core/src/Kokkos_Macros.hpp

```diff
@@ -357,6 +357,10 @@
 #define KOKKOS_IMPL_DEVICE_FUNCTION
 #endif
 
+#if !defined(KOKKOS_IMPL_EXTERNAL_FUNCTION)
+#define KOKKOS_IMPL_EXTERNAL_FUNCTION KOKKOS_IMPL_FUNCTION
```

Comment:
I think we should consider including the `extern` storage class specifier in that macro definition.

Reply:
If we need to use in the definition as well, it seems strange to use `extern` but that would work  as well. What case do you have in mind were `extern` would be required and not redundant?

Reply:
How does that seem strange?  It is literally in the macro's name.

Reply:
Yes, `extern` is in the macro's name but it doesn't do the same as `extern` in "regular" C++ code for free functions (which is redundant anyway). It's possible to use `extern` also for the definition of a function but it doesn't carry the intent that the function is defined in a different translation.

All in all, I lean slightly towards not adding `extern`.

Reply:
should we call it relocatable then? I.e. KOKKOS_RELOCTABLE_FUNCTION?

Reply:
I am in favor of renaming the macro to `KOKKOS_RELOCATABLE_FUNCTION`.